File: bot-api/dotnet/Robocode.TankRoyale.BotApi/src/internal/BotEventHandlers.cs
using System;
using System.Collections.Generic;
using Robocode.TankRoyale.BotApi.Events;

namespace Robocode.TankRoyale.BotApi.Internal;

/// <summary>
/// Class used for bot event handlers on the public API that are may or may not be triggered bot event queue, and might
/// not be handled immediately by the bot logic.
/// </summary>
sealed class BotEventHandlers
{
    internal readonly EventHandler<ConnectedEvent> OnConnected = new();
    internal readonly EventHandler<DisconnectedEvent> OnDisconnected = new();
    internal readonly EventHandler<ConnectionErrorEvent> OnConnectionError = new();

    internal readonly EventHandler<GameStartedEvent> OnGameStarted = new();
    internal readonly EventHandler<GameEndedEvent> OnGameEnded = new();
    internal readonly EventHandler<GameAbortedEvent> OnGameAborted = new();
    internal readonly EventHandler<RoundStartedEvent> OnRoundStarted = new();
    internal readonly EventHandler<RoundEndedEvent> OnRoundEnded = new();

    private readonly EventHandler<TickEvent> OnTick = new();
    internal readonly EventHandler<SkippedTurnEvent> OnSkippedTurn = new();
    private readonly EventHandler<DeathEvent> OnDeath = new();
    private readonly EventHandler<BotDeathEvent> OnBotDeath = new();
    private readonly EventHandler<HitBotEvent> OnHitBot = new();
    private readonly EventHandler<HitWallEvent> OnHitWall = new();
    private readonly EventHandler<BulletFiredEvent> OnBulletFired = new();
    private readonly EventHandler<HitByBulletEvent> OnHitByBullet = new();
    private readonly EventHandler<BulletHitBotEvent> OnBulletHit = new();
    private readonly EventHandler<BulletHitBulletEvent> OnBulletHitBullet = new();
    private readonly EventHandler<BulletHitWallEvent> OnBulletHitWall = new();
    private readonly EventHandler<ScannedBotEvent> OnScannedBot = new();
    private readonly EventHandler<WonRoundEvent> OnWonRound = new();
    private readonly EventHandler<CustomEvent> OnCustomEvent = new();
    private readonly EventHandler<TeamMessageEvent> OnTeamMessage = new();

    private readonly Dictionary<Type, Action<IEvent>> _eventHandlers = new();


    internal BotEventHandlers(IBaseBot baseBot)
    {
        InitializeEventHandlers();
        SubscribeToEventHandlers(baseBot);
    }

    private void InitializeEventHandlers()
    {
        _eventHandlers[typeof(ConnectedEvent)] = e => OnConnected.Publish((ConnectedEvent)e);
        _eventHandlers[typeof(DisconnectedEvent)] = e => OnDisconnected.Publish((DisconnectedEvent)e);
        _eventHandlers[typeof(ConnectionErrorEvent)] = e => OnConnectionError.Publish((ConnectionErrorEvent)e);

        _eventHandlers[typeof(GameStartedEvent)] = e => OnGameStarted.Publish((GameStartedEvent)e);
        _eventHandlers[typeof(GameEndedEvent)] = e => OnGameEnded.Publish((GameEndedEvent)e);
        _eventHandlers[typeof(GameAbortedEvent)] = e => OnGameAborted.Publish((GameAbortedEvent)e);
        _eventHandlers[typeof(RoundStartedEvent)] = e => OnRoundStarted.Publish((RoundStartedEvent)e);
        _eventHandlers[typeof(RoundEndedEvent)] = e => OnRoundEnded.Publish((RoundEndedEvent)e);

        _eventHandlers[typeof(TickEvent)] = e => OnTick.Publish((TickEvent)e);
        _eventHandlers[typeof(SkippedTurnEvent)] = e => OnSkippedTurn.Publish((SkippedTurnEvent)e);
        _eventHandlers[typeof(DeathEvent)] = e => OnDeath.Publish((DeathEvent)e);
        _eventHandlers[typeof(BotDeathEvent)] = e => OnBotDeath.Publish((BotDeathEvent)e);
        _eventHandlers[typeof(HitBotEvent)] = e => OnHitBot.Publish((HitBotEvent)e);
        _eventHandlers[typeof(HitWallEvent)] = e => OnHitWall.Publish((HitWallEvent)e);
        _eventHandlers[typeof(BulletFiredEvent)] = e => OnBulletFired.Publish((BulletFiredEvent)e);
        _eventHandlers[typeof(HitByBulletEvent)] = e => OnHitByBullet.Publish((HitByBulletEvent)e);
        _eventHandlers[typeof(BulletHitBotEvent)] = e => OnBulletHit.Publish((BulletHitBotEvent)e);
        _eventHandlers[typeof(BulletHitBulletEvent)] = e => OnBulletHitBullet.Publish((BulletHitBulletEvent)e);
        _eventHandlers[typeof(BulletHitWallEvent)] = e => OnBulletHitWall.Publish((BulletHitWallEvent)e);
        _eventHandlers[typeof(ScannedBotEvent)] = e => OnScannedBot.Publish((ScannedBotEvent)e);
        _eventHandlers[typeof(WonRoundEvent)] = e => OnWonRound.Publish((WonRoundEvent)e);
        _eventHandlers[typeof(CustomEvent)] = e => OnCustomEvent.Publish((CustomEvent)e);
        _eventHandlers[typeof(TeamMessageEvent)] = e => OnTeamMessage.Publish((TeamMessageEvent)e);
    }

    private void SubscribeToEventHandlers(IBaseBot baseBot)
    {
        OnConnected.Subscribe(baseBot.OnConnected);
        OnDisconnected.Subscribe(baseBot.OnDisconnected);
        OnConnectionError.Subscribe(baseBot.OnConnectionError);

        OnGameStarted.Subscribe(baseBot.OnGameStarted);
        OnGameEnded.Subscribe(baseBot.OnGameEnded);
        OnRoundStarted.Subscribe(baseBot.OnRoundStarted);
        OnRoundEnded.Subscribe(baseBot.OnRoundEnded);

        OnTick.Subscribe(baseBot.OnTick);
        OnSkippedTurn.Subscribe(baseBot.OnSkippedTurn);
        OnDeath.Subscribe(baseBot.OnDeath);
        OnBotDeath.Subscribe(baseBot.OnBotDeath);
        OnHitBot.Subscribe(baseBot.OnHitBot);
        OnHitWall.Subscribe(baseBot.OnHitWall);
        OnBulletFired.Subscribe(baseBot.OnBulletFired);
        OnHitByBullet.Subscribe(baseBot.OnHitByBullet);
        OnBulletHit.Subscribe(baseBot.OnBulletHit);
        OnBulletHitBullet.Subscribe(baseBot.OnBulletHitBullet);
        OnBulletHitWall.Subscribe(baseBot.OnBulletHitWall);
        OnScannedBot.Subscribe(baseBot.OnScannedBot);
        OnWonRound.Subscribe(baseBot.OnWonRound);
        OnCustomEvent.Subscribe(baseBot.OnCustomEvent);
        OnTeamMessage.Subscribe(baseBot.OnTeamMessage);
    }

    internal void FireEvent(BotEvent botEvent)
    {
        if (_eventHandlers.TryGetValue(botEvent.GetType(), out var action))
        {
            action.Invoke(botEvent); // calls Publish(botEvent)
        }
        else
        {
            throw new InvalidOperationException("Unhandled event: " + botEvent);
        }
    }

    // Virtual (fake) event:
    internal sealed class GameAbortedEvent : IEvent
    {
    }
}

File: bot-api/java/src/main/java/dev/robocode/tankroyale/botapi/internal/BotEventHandlers.java
package dev.robocode.tankroyale.botapi.internal;

import dev.robocode.tankroyale.botapi.IBaseBot;
import dev.robocode.tankroyale.botapi.events.*;

import java.util.HashMap;
import java.util.Map;

/**
 * Class used for bot event handlers on the public API that are may or may not be triggered bot event queue, and might
 * not be handled immediately by the bot logic.
 */
final class BotEventHandlers {

    final EventHandler<ConnectedEvent> onConnected = new EventHandler<>();
    final EventHandler<DisconnectedEvent> onDisconnected = new EventHandler<>();
    final EventHandler<ConnectionErrorEvent> onConnectionError = new EventHandler<>();

    final EventHandler<GameStartedEvent> onGameStarted = new EventHandler<>();
    final EventHandler<GameEndedEvent> onGameEnded = new EventHandler<>();
    final EventHandler<GameAbortedEvent> onGameAborted = new EventHandler<>();
    final EventHandler<RoundStartedEvent> onRoundStarted = new EventHandler<>();
    final EventHandler<RoundEndedEvent> onRoundEnded = new EventHandler<>();

    final EventHandler<TickEvent> onTick = new EventHandler<>();
    final EventHandler<SkippedTurnEvent> onSkippedTurn = new EventHandler<>();
    final EventHandler<DeathEvent> onDeath = new EventHandler<>();
    final EventHandler<BotDeathEvent> onBotDeath = new EventHandler<>();
    final EventHandler<HitBotEvent> onHitBot = new EventHandler<>();
    final EventHandler<HitWallEvent> onHitWall = new EventHandler<>();
    final EventHandler<BulletFiredEvent> onBulletFired = new EventHandler<>();
    final EventHandler<HitByBulletEvent> onHitByBullet = new EventHandler<>();
    final EventHandler<BulletHitBotEvent> onBulletHit = new EventHandler<>();
    final EventHandler<BulletHitBulletEvent> onBulletHitBullet = new EventHandler<>();
    final EventHandler<BulletHitWallEvent> onBulletHitWall = new EventHandler<>();
    final EventHandler<ScannedBotEvent> onScannedBot = new EventHandler<>();
    final EventHandler<WonRoundEvent> onWonRound = new EventHandler<>();
    final EventHandler<CustomEvent> onCustomEvent = new EventHandler<>();
    final EventHandler<TeamMessageEvent> onTeamMessage = new EventHandler<>();

    private final Map<Class<? extends IEvent>, EventHandler<? extends IEvent>> eventHandlerMap = new HashMap<>();


    BotEventHandlers(IBaseBot baseBot) {
        initializeEventHandlers();
        subscribeToEventHandlers(baseBot);
    }

    private void initializeEventHandlers() {
        eventHandlerMap.put(ConnectedEvent.class, onConnected);
        eventHandlerMap.put(DisconnectedEvent.class, onDisconnected);
        eventHandlerMap.put(ConnectionErrorEvent.class, onConnectionError);

        eventHandlerMap.put(GameStartedEvent.class, onGameStarted);
        eventHandlerMap.put(GameEndedEvent.class, onGameEnded);
        eventHandlerMap.put(GameAbortedEvent.class, onGameAborted);
        eventHandlerMap.put(RoundStartedEvent.class, onRoundStarted);
        eventHandlerMap.put(RoundEndedEvent.class, onRoundEnded);

        eventHandlerMap.put(TickEvent.class, onTick);
        eventHandlerMap.put(SkippedTurnEvent.class, onSkippedTurn);
        eventHandlerMap.put(DeathEvent.class, onDeath);
        eventHandlerMap.put(BotDeathEvent.class, onBotDeath);
        eventHandlerMap.put(HitBotEvent.class, onHitBot);
        eventHandlerMap.put(HitWallEvent.class, onHitWall);
        eventHandlerMap.put(BulletFiredEvent.class, onBulletFired);
        eventHandlerMap.put(HitByBulletEvent.class, onHitByBullet);
        eventHandlerMap.put(BulletHitBotEvent.class, onBulletHit);
        eventHandlerMap.put(BulletHitBulletEvent.class, onBulletHitBullet);
        eventHandlerMap.put(BulletHitWallEvent.class, onBulletHitWall);
        eventHandlerMap.put(ScannedBotEvent.class, onScannedBot);
        eventHandlerMap.put(WonRoundEvent.class, onWonRound);
        eventHandlerMap.put(CustomEvent.class, onCustomEvent);
        eventHandlerMap.put(TeamMessageEvent.class, onTeamMessage);
    }

    private void subscribeToEventHandlers(IBaseBot baseBot) {
        onConnected.subscribe(baseBot::onConnected);
        onDisconnected.subscribe(baseBot::onDisconnected);
        onConnectionError.subscribe(baseBot::onConnectionError);

        onGameStarted.subscribe(baseBot::onGameStarted);
        onGameEnded.subscribe(baseBot::onGameEnded);
        onRoundStarted.subscribe(baseBot::onRoundStarted);
        onRoundEnded.subscribe(baseBot::onRoundEnded);

        onTick.subscribe(baseBot::onTick);
        onSkippedTurn.subscribe(baseBot::onSkippedTurn);
        onDeath.subscribe(baseBot::onDeath);
        onBotDeath.subscribe(baseBot::onBotDeath);
        onHitBot.subscribe(baseBot::onHitBot);
        onHitWall.subscribe(baseBot::onHitWall);
        onBulletFired.subscribe(baseBot::onBulletFired);
        onHitByBullet.subscribe(baseBot::onHitByBullet);
        onBulletHit.subscribe(baseBot::onBulletHit);
        onBulletHitBullet.subscribe(baseBot::onBulletHitBullet);
        onBulletHitWall.subscribe(baseBot::onBulletHitWall);
        onScannedBot.subscribe(baseBot::onScannedBot);
        onWonRound.subscribe(baseBot::onWonRound);
        onCustomEvent.subscribe(baseBot::onCustomEvent);
        onTeamMessage.subscribe(baseBot::onTeamMessage);
    }

    @SuppressWarnings("unchecked")
    void fireEvent(BotEvent event) {
        var handler = (EventHandler<IEvent>) eventHandlerMap.get(event.getClass());
        if (handler != null) {
            handler.publish(event);
        } else {
            throw new IllegalStateException("Unhandled event type: " + event);
        }
    }

    // Virtual (fake) event:
    private static final class GameAbortedEvent implements IEvent {
    }
}